tests/prague/eip7702_set_code_tx/test_set_code_txs.py

```diff
@@ -0,0 +1,2482 @@
+    Tests use of set-code transactions from [EIP-7702: Set EOA account code for one transaction](https://eips.ethereum.org/EIPS/eip-7702).
+"""  # noqa: E501
+
+from enum import Enum
+from hashlib import sha256
+from itertools import count
+from typing import List
+
+import pytest
+from ethereum.crypto.hash import keccak256
+
+from ethereum_test_tools import (
+    AccessList,
+    Account,
+    Address,
+    Alloc,
+    AuthorizationTuple,
+    Block,
+    BlockchainTestFiller,
+    Bytecode,
+    CodeGasMeasure,
+    Conditional,
+    Environment,
+    EVMCodeType,
+    Hash,
+    Initcode,
+)
+from ethereum_test_tools import Macros as Om
+from ethereum_test_tools import Opcodes as Op
+from ethereum_test_tools import (
+    StateTestFiller,
+    Storage,
+    Transaction,
+    TransactionException,
+    add_kzg_version,
+    call_return_code,
+    compute_create_address,
+)
+from ethereum_test_tools.eof.v1 import Container, Section
+
+from ...cancun.eip4844_blobs.spec import Spec as Spec4844
+from ..eip6110_deposits.helpers import DepositRequest
+from ..eip7002_el_triggerable_withdrawals.helpers import WithdrawalRequest
+from ..eip7251_consolidations.helpers import ConsolidationRequest
+from .helpers import AddressType
+from .spec import Spec, ref_spec_7702
+
+REFERENCE_SPEC_GIT_PATH = ref_spec_7702.git_path
+REFERENCE_SPEC_VERSION = ref_spec_7702.version
+
+pytestmark = pytest.mark.valid_from("Prague")
+
+auth_account_start_balance = 0
+
+
+@pytest.mark.parametrize(
+    "tx_value",
+    [0, 1],
+)
+@pytest.mark.parametrize(
+    "suffix,succeeds",
+    [
+        pytest.param(Op.STOP, True, id="stop"),
+        pytest.param(Op.RETURN(0, 0), True, id="return"),
+        pytest.param(Op.REVERT, False, id="revert"),
+        pytest.param(Op.INVALID, False, id="invalid"),
+        pytest.param(Om.OOG, False, id="out-of-gas"),
+    ],
+)
+def test_self_sponsored_set_code(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    suffix: Bytecode,
+    succeeds: bool,
+    tx_value: int,
+):
+    """
+    Test the executing a self-sponsored set-code transaction.
+
+    The transaction is sent to the sender, and the sender is the signer of the only authorization
+    tuple in the authorization list.
+
+    The authorization tuple has a nonce of 1 because the self-sponsored transaction increases the
+    nonce of the sender from zero to one first.
+
+    The expected nonce at the end of the transaction is 2.
+    """
+    storage = Storage()
+    sender = pre.fund_eoa()
+
+    set_code = (
+        Op.SSTORE(storage.store_next(sender), Op.ORIGIN)
+        + Op.SSTORE(storage.store_next(sender), Op.CALLER)
+        + Op.SSTORE(storage.store_next(tx_value), Op.CALLVALUE)
+        + suffix
+    )
+    set_code_to_address = pre.deploy_contract(
+        set_code,
+    )
+
+    tx = Transaction(
+        gas_limit=10_000_000,
+        to=sender,
+        value=tx_value,
+        authorization_list=[
+            AuthorizationTuple(
```

Comment:
Here (and I think in many other tests) Transaction's chain_id is 1 by default, and authorization chain_id is 0, so it should not be valid?

Reply:
It's valid because the spec allows 0 chain_id as a "global" id, so it's valid on any network.

Reply:
ah right, I missed that